### ISSUE TITLE
Can't use NuGet packages with 4 (or more) dot version number

### ISSUE THREAD

**Issue by KuraiAndras** [xoofx/UnityNuGet#55]
I tried out UnityNuGet, by running it locally, and adding dependencies for [OPCFoundation.NetStandard.Opc.Ua.Client](https://www.nuget.org/packages/OPCFoundation.NetStandard.Opc.Ua.Client/). This package has versions with 4 parts. For example the latest version is: 1.4.367.41. When trying to import it with the package manager, I get an error:
```
[Package Manager Window] Cannot perform upm operation: Unable to add package [org.nuget.opcfoundation.netstandard.opc.ua.client@1.4.367-41]:
  Package org.nuget.opcfoundation.netstandard.opc.ua.client@1.4.367-41 has invalid dependencies or related test packages:
    org.nuget.opcfoundation.netstandard.opc.ua.configuration (dependency): Version '1.4.367.41' is invalid. Expected a 'SemVer' compatible value.
    org.nuget.opcfoundation.netstandard.opc.ua.core (dependency): Version '1.4.367.41' is invalid. Expected a 'SemVer' compatible value. [NotFound].
UnityEditor.EditorApplication:Internal_CallUpdateFunctions ()
```

Edit: Also i could not see any warning or error coming from the server.

**Comment by xoofx** [comment]
Yes, It is a known issue. NPM doesn't support 4 digits. Could you try to modify your local UnityNuGet and experiment with e.g `1.4.367+41` instead to see if it is working (`+` is supposed for the build part, but I don't know if Unity NPM supports it)

**Comment by xoofx** [comment]
Oh, it's actually a bug in the dependencies, let me see If I can fix it

**Comment by xoofx** [comment]
This should be fixed by commit 628a423 but I haven't added any tests so if you can try this out from latest master, thanks

**Comment by KuraiAndras** [comment]
Thanks, I will try it out soon

**Comment by KuraiAndras** [comment]
I don't know if this is related to the fix or not, but I still can't add the package.

`OPCFoundation.NetStandard.Opc.Ua.Core` has a dependency on `System.Data.Common` and `System.Net.NameResolution `. I added these to the registry:
```json
  "System.Data.Common": {
    "listed": true,
    "version": "4.3.0"
  },
  "System.Net.NameResolution": {
    "listed": true,
    "version": "4.3.0"
  }
```
These are listed in the `DotNetHelper.NetStandard21Assemblies`, and they don't provide .net standrad 2.0 assemblies on nuget.
In the logs a warning is shown as expected:
```
warn: NuGet[0]
      The package `System.Data.Common.4.3.0` doesn't support `netstandard2.1,netstandard2.0`
...
warn: NuGet[0]
      The package `System.Net.NameResolution.4.3.0` doesn't support `netstandard2.1,netstandard2.0`
```

When I try to install with the package manager I get an error:

```
[Package Manager Window] Cannot perform upm operation: Unable to add package [org.nuget.opcfoundation.netstandard.opc.ua.client@1.4.367-41]:
  Dependency org.nuget.opcfoundation.netstandard.opc.ua.core@1.4.367-41 has invalid dependencies or related test packages:
    org.nuget.system.data.common (dependency): Package [org.nuget.system.data.common@4.3.0] cannot be found
    org.nuget.system.net.nameresolution (dependency): Package [org.nuget.system.net.nameresolution@4.3.0] cannot be found [NotFound].
UnityEditor.EditorApplication:Internal_CallUpdateFunctions ()
```

After this I tried ignoring the System* packages:
```json
  "System.Data.Common": {
    "ignore": true
  },
  "System.Net.NameResolution": {
    "ignore": true
  }
```

But then I get another error:

```
[Package Manager Window] Cannot perform upm operation: Unable to add package [org.nuget.opcfoundation.netstandard.opc.ua.client@1.4.367-41]:
  One or more dependencies could not be added to the local file system:
    org.nuget.system.numerics.vectors: Tarball download failed after [1] attempts. Downloaded [61892] bytes, expected [61892]:
      shasum check failed for C:\Users\kurai.andras.AUTSOFT\AppData\Local\Unity\cache\npm\localhost_5321\.tmp78248f0GH71X7nX5\tarball\package.tgz
      Expected: dbd9ea0bf6ee1e1753d63e2f36a121594f5bbdd2
      Actual:   ae41184e0dd71ebe969480d3ccba2f5bb85eab75 [NotFound].
UnityEditor.EditorApplication:Internal_CallUpdateFunctions ()
```

So then I tried ignoring the System.Numerics.Vectors package, and that let me add the package, but I still have issues, I think with BouncyCastle.Portable.

I will investigate further to find the underlying issue.

BTW I was able to use theses packages, by manually adding the DLLs, so this should work.

One hint I have is that BouncyCastle.Portable has a DLL named BouncyCastle.Crypto, and the inspector shows that it has a missing Define Constraints:
![image](https://user-images.githubusercontent.com/19240945/136009510-e908f91c-a23d-4573-9b78-889e89b7e68b.png)

**Comment by bdovaz** [comment]
@KuraiAndras I have seen before the problem you are facing.

It's because some packages have dependencies on packages that are already on .NET Standard 2.0+ runtime so there shouldn't be there defined as dependencies.

For example, the following dependencies have no updates since 2016:

https://www.nuget.org/packages/System.Data.Common/
https://www.nuget.org/packages/System.Net.NameResolution/

In some cases I have created PR's to fix this problem that it's only "a problem" with Unity + UnityNuget.

**Comment by KuraiAndras** [comment]
Thanks @bdovaz I'll try it soon

**Comment by xoofx** [comment]
With which Unity version are you testing this?

**Comment by KuraiAndras** [comment]
2021.1.22

**Comment by bdovaz** [comment]
@KuraiAndras that Tarball problem I think that it's a discrepancy on unityNuget registry vs your local machine cache. If you clean your local cache I think that it should be fixed.

**Comment by xoofx** [comment]
> @KuraiAndras that Tarball problem I think that it's a discrepancy on unityNuget registry vs your local machine cache. If you clean your local cache I think that it should be fixed.

Yes, I forgot to mention, the old cache will cause many conflicts and needs to be deleted manually

**Comment by xoofx** [comment]
> One hint I have is that BouncyCastle.Portable has a DLL named BouncyCastle.Crypto, and the inspector shows that it has a missing Define Constraints:

Also the missing define constraints might be a problem with the existing code that doesn't add a define constraint if the package is 2.0. But it might require to remove the line entirely in the yml file.

**Comment by KuraiAndras** [comment]
Okay, so I cleared the npm caches, the unity_packages cache, and the Library folder in the unity project. @bdovaz I removed the `System.Data.Common` and the `System.Net.NameResolution` packages from the registry, but then the server shows an error, and when adding the package from unity I get an error, because it can't resolve them
```
[Package Manager Window] Cannot perform upm operation: Unable to add package [org.nuget.opcfoundation.netstandard.opc.ua.client@1.4.367-41]:
  Dependency org.nuget.opcfoundation.netstandard.opc.ua.core@1.4.367-41 has invalid dependencies or related test packages:
    org.nuget.system.data.common (dependency): Package [org.nuget.system.data.common@4.3.0] cannot be found
    org.nuget.system.net.nameresolution (dependency): Package [org.nuget.system.net.nameresolution@4.3.0] cannot be found [NotFound].
UnityEditor.EditorApplication:Internal_CallUpdateFunctions ()
```

After clearing the caches I also tried adding them as ignored in the registry, then adding them worked 🎉, but now I can't reference anything from them.

```
error CS0246: The type or namespace name 'Session' could not be found (are you missing a using directive or an assembly reference?)
```

I can see them in the `Project/Packages`:
![image](https://user-images.githubusercontent.com/19240945/136043842-b80940da-2a83-4bc6-b472-feaab093f17e.png)

I can't see the DLLs in the generated csproj files. Also, when I inspect the DLLs in the editor, Unity shows a warning(error?) for the  Define Constraints:

![image](https://user-images.githubusercontent.com/19240945/136045190-6377021f-a7ba-4207-a3fc-2b6bb7acb2a0.png)

It is also present in MoreLinq
![image](https://user-images.githubusercontent.com/19240945/136045262-8de68b47-3a63-4447-a208-aabfa7bd4d3d.png)

I tired manually editing the `.meta` files for the DLLs, but as soon as I change one unity reimports it with the original content.
